### ISSUE TITLE
Superset does not support Arabic language or RTL layout

### ISSUE THREAD

**Issue by ca5p3r** [apache/superset#25258]
## Screenshots

![image](https://github.com/apache/superset/assets/39639769/ed307cb8-30e5-4875-a2e9-1fcccdbec5f2)

## Description

Superset is a powerful data visualization tool, but it does not currently support the Arabic language or the RTL layout. This makes it difficult for users in Arabic-speaking countries to create Superset dashboards in Arabic.

The following are some of the problems that users in Arabic-speaking countries experience when using Superset:

- There's no support for Arabic Language in superset UI
- The UI elements are not aligned correctly in the RTL layout
- The keyboard shortcuts are not mapped to the Arabic keyboard layout
 
I would like to request that Superset be updated to support the Arabic language and the RTL layout. This would make Superset a more accessible tool for users in Arabic countries.

**Comment by omar-nelc** [comment]
Hello maintainers,

I'm Omar Al-Ithawi and I'm a Core Contributor in the @openedx project which is also built in Python and React.

In 2014, the Open edX  project -- a Learning management system -- had no Arabic or Right-to-Left support and often times lacked uft-8 support on some modules.

Many contributors and myself have worked on this issue until Unicode, Arabic and Right-to-Left has been supported.

I'm planning to do the same for superset. I'm assuming Unicode is supported by Python 3.9, so we need to support Arabic and RTL styles.

For Arabic, we're planning to add translations because it seems to be fairly doable. However, we're reviewing the status of RTL support at the moment to see how can we add it.

**Comment by alimahwer** [comment]
Hi Omar, I have experience translating some open-source projects into Arabic. So, I will be glad to contribute to the Arabization of Superset. However, We are still waiting for who will take the initiative of adding RTL orientation to Superset. Hope we together can cooperate on this task.

**Comment by omar-nelc** [comment]
@alimahwer Please feel free to start translating the project even though it lacks RTL support.

I think the two problems can be fixed in parallel, so if you have the capacity to translate it, please do so and tag me for a review if needed.

**Comment by bmtcril** [comment]
Hi all, has there been any movement on this issue? I haven't been able to find anything related.

**Comment by rusackas** [comment]
Nope, I think this is in need of volunteers. I'm tempted to close this out as an Issue, since it's not a bug, but rather a feature request. Normally I would, but I'm hopeful someone on this thread might pick it up. If not, we should move it to a Discussion or close it.

**Comment by bmtcril** [comment]
@omar-nelc is one of our contributors so I'll round up with him and see if we can help push this forward, we'd love to have it!

**Comment by OmarIthawi** [comment]
@rusackas and @bmtcril thanks for keeping this issue active.

The lowest hanging fruit is to run [rtlcss](https://rtlcss.com/index.html) on the CSS files and it would generate two files: 

 - Left to Right CSS file (original)
 - Right to Left CSS file (generated)

This framework is flexible and works with many other frameworks and systems.

The other part is to ensure the UI loads the correct file depending on the chosen user language.

Would the Superset team be willing to add such build steps? Whether we contribute it or the core team does that.

Suppose I'd add those, would you mind give us pointers on where to add those build steps, introduce a new dependency and edit the CSS files when necessary.

There will be some UI issues after we and it's manual frontend work to fix those. From system to system it varies, but it's a doable work in general.

**Comment by rusackas** [comment]
> Whether we contribute it or the core team does that.

We're open to contributions! You're right that `<styled>` (or any other Emotion-based theming methods) would need to support this. We're moving toward that library and away from LESS/CSS files, so anything based on _only_ those files will have less and less impact as we move away from them.

I don't know of anyone on the PMC/Committer roster who has this on their roadmap, so we'd have to lean on other contributors from this thread to take action here :) Thanks in advance!

**Comment by OmarIthawi** [comment]
Thanks @rusackas that makes it clear.

I'll see if my team is up to take a stab on this.

The initial suggested approach is to provide a custom `styled` with the following logic:

```
// Pseudo-code here
const bidiStyled = (args...) => {
    args.styles = args.styles.replace('margin-left', 'margin-right');
    return styled(args...);
}

const Section = bidiStyled.section`
  margin-left: 20px;
  padding-top: 10px;
`
```

In an RTL document, it would cause `margin-right` to be used.

An alternative is switch hard-coded direction properties like `padding-left` to context-dependent  `padding-block-start` and similar properties. What's your opinion on that?

 - https://developer.mozilla.org/en-US/docs/Web/CSS/padding-inline-start
 - https://developer.mozilla.org/en-US/docs/Web/CSS/margin-block-start

Applying those properties would allow (ideally) a single switch `<html dir="rtl">` to flip the page direction and the rest is the browsers-work:

I've checked browser support and it seems to have good support 96%+:

 - https://caniuse.com/mdn-css_properties_margin-block-start

Luckily, flex UI and many other CSS properties automatically adjust when `<html dir=rtl>` is used as shown in the screenshot below -- it's not too broken, which is a pleasant surprise!!

![image](https://github.com/apache/superset/assets/645156/99290809-2097-4544-85f3-de8d80607082)

**Comment by rusackas** [comment]
I love the idea of switching to `{padding/margin}-block-{start/end}` if that unblocks this. My two questions:

• Is there a listing rule that can protect against people using the more common/obvious padding/margin attributes?
• Would/could a listing rule actually _fix_ this?
• How many people use IE11, and is it OK to officially drop support for this? I would think everyone is using Edge by now... I don't see any IE users in the Superset website's stats, but we don't really know this stat for Superset itself.

Pinging a few others here who might have opinions on the matter: @kgabryje @mistercrunch @michael-s-molina

**Comment by OmarIthawi** [comment]
> I love the idea of switching to `{padding/margin}-block-{start/end}` if that unblocks this. My two questions:

Great!

> • Is there a listing rule that can protect against people using the more common/obvious padding/margin attributes? 

I'd think not. This use case is very specific React JSX and Styled, but I presume we could fallback to basic `grepping` and develop a rather naive string-checking approach and it would work. I hope there's a better code aware tool, but once we start developing it'll be clearer.

> • Would/could a listing rule actually _fix_ this? 

Usually yes. But it'll be annoying to both first-time and occasional contributors.

> • How many people use IE11, and is it OK to officially drop support for this? I would think everyone is using Edge by now... I don't see any IE users in the Superset website's stats, but we don't really know this stat for Superset itself.

I can't say much about that.

**Comment by michael-s-molina** [comment]
My two cents:
- RLT is a property of the [theme object](https://mui.com/material-ui/customization/right-to-left/)
- It's not only spacing but also [icon mirroring and representing the passage of time correctly](https://m2.material.io/design/usability/bidirectionality.html#mirroring-layout)

**Comment by OmarIthawi** [comment]
Status: 

 - Translations is on the way, with no specific deadline yet.
 - RTL work is still not started.

**Comment by CodeWithEmad** [comment]
Hey @OmarIthawi, just checking in to see if there have been any updates on this over the past few months.

**Comment by OmarIthawi** [comment]
@CodeWithEmad the translations is ready, and we're pushing it to Superset soon. RTL work is still needed.

**Comment by OmarIthawi** [comment]
@alimahwer The arabic translations has been mostly completed by my colleauge @abdilra7eem in

 - https://github.com/apache/superset/pull/29586

If you want to take another pass and to see if more translation is needed please go ahead.

However, the critical need at the moment is RTL. No one is working on RTL at the moment as far as I know.

**Comment by CodeWithEmad** [comment]
Hi @OmarIthawi. I haven't done anything in the RTL area in the superset project, but I'll be glad to help.

**Comment by OmarIthawi** [comment]
Same here @CodeWithEmad, but glad you're willing to help.

Please review the comments in this https://github.com/apache/superset/issues/25258 issue, that's all what we have at the moment.

**Comment by OmarIthawi** [comment]
@mayankeagle no work has been done on RTL support as far as I know. Only Arabic translations on the LTR User Interface.

**Comment by mayankeagle** [comment]
> @mayankeagle no work has been done on RTL support as far as I know. Only Arabic translations on the LTR User Interface.

Hi @OmarIthawi Ok, thanks. And are the Arabic translations supported on the latest GA release version of Superset, under Apache 2.0 license?

**Comment by OmarIthawi** [comment]
@mayankeagle, no. The Arabic isn't released yet: https://github.com/apache/superset/releases

It's only available in the main branch.

**Comment by OmarIthawi** [comment]
here you go @mayankeagle https://github.com/apache/superset/pull/29586

**Comment by CodeWithEmad** [comment]
TIL that Superset uses [Flask-AppBuilder](https://github.com/dpgaspar/Flask-AppBuilder)
and to have full support for the Arabic language, you need to add translations to that project too.

**Comment by OmarIthawi** [comment]
Good note, thanks @CodeWithEmad! I think this one has to wait a bit. Superset at Open edX is not very well implemented at the moment and we're not actively using it as we should so translation have stalled.

**Comment by rusackas** [comment]
Anyone still interested in tackling this? It's been untouched for nearly half a year so I can either close it as stale or move it to an Ideas Discussion.

**Comment by supapo** [comment]
> Anyone still interested in tackling this? It's been untouched for nearly half a year so I can either close it as stale or move it to an Ideas Discussion.

Yes it is very important

**Comment by OmarIthawi** [comment]
Arabic is merged, but RTL still needs contributors

**Comment by CodeWithEmad** [comment]
[Persian PR](https://github.com/apache/superset/pull/29580) has still been waiting for a review since last July! :)

**Comment by alrehaili** [comment]
is it solved ?

**Comment by OmarIthawi** [comment]
Not @alrehaili, but you're welcome to contribute RTL fixes if you'd like :)

**Comment by msyavuz** [comment]
Ant Design has an option for changing text direction in the ConfigProvider. It might help to see if it works in superset out of the box, if anyones interested in taking this on

**Comment by ceenayekta** [comment]
@michael-s-molina @omar-nelc RTL Support added on #34629 .
I've just fixed some common UI alignment issues, and used the new ThemeController to set direction to `'rtl'` as we change the language.

I'm not exactly sure if we should revert icons.

**Comment by OmarIthawi** [comment]
Great work @ceenayekta! Thanks for your contribution.

Yes, some icons need to be reverted, esp. the ones with directional Arrows:

 - ↩️ 
 - ◀️ 
 - ➡️ 

Anything with a global fixed direction should be flipped. Examples:

 - Social media icons shouldn't be reversed
 - Any logo :octocat: 
 - Glob icons 🌐 
 - Maps 🗺️ 
 - Flags 🇿🇦 

Some icons are okay whether flipped or not:

 - 👮🏼‍♂️ 
 - 🚗 
 - 👍🏼 

And so on

**Comment by rusackas** [comment]
Closing this since it's mostly (?) done, but we're open to PRs, or Discussions/Issues on any remaining follow-up items that need to be tended to.
